File: Python/descriptives/stats_descriptives_QQ.py
#!/usr/bin/env python
# coding: utf-8

# # COURSE: Master statistics and machine learning: Intuition, Math, code
# ##### COURSE URL: udemy.com/course/statsml_x/?couponCode=202304 
# ## SECTION: Descriptive statistics
# ### VIDEO: QQ plots
# #### TEACHER: Mike X Cohen, sincxpress.com

# In[ ]:


# import libraries
import matplotlib.pyplot as plt
import numpy as np
import scipy.stats as stats


# In[ ]:


## generate data

n    = 1000
data = np.random.randn(n)
# data = np.exp( np.random.randn(n)*.8 ) # log-norm distribution

# theoretical normal distribution given N
x = np.linspace(-4,4,10001)
theonorm = stats.norm.pdf(x)
theonorm = theonorm/max(theonorm)

# plot histograms on top of each other
yy,xx = np.histogram(data,40)
yy = yy/max(yy)
xx = (xx[:-1]+xx[1:])/2

plt.plot(xx,yy,label='Empirical')
plt.plot(x,theonorm,label='Theoretical')
plt.legend()
plt.show()


# In[ ]:


## create a QQ plot

zSortData  = np.sort(stats.zscore(data))
sortNormal = stats.norm.ppf(np.linspace(0,1,n))

# QQ plot is theory vs reality
plt.plot(sortNormal,zSortData,'o')

# set axes to be equal
xL,xR = plt.xlim()
yL,yR = plt.ylim()
lims  = [ np.min([xL,xR,yL,yR]),np.max([xL,xR,yL,yR]) ]
plt.xlim(lims)
plt.ylim(lims)

# draw red comparison line
plt.plot(lims,lims)

plt.xlabel('Theoretical normal')
plt.ylabel('Observed data')
plt.title('QQ plot')
plt.axis('square')
plt.show()


# In[ ]:


## Python solution

x = stats.probplot(data,plot=plt)

File: MATLAB/descriptives/stats_descriptives_QQ.m
%%
%   COURSE: Master statistics and machine learning: intuition, math, code										
%      URL: udemy.com/course/statsml_x/?couponCode=202304 
% 
%  SECTION: Descriptive statistics
%    VIDEO: QQ plots
% 
%  TEACHER: Mike X Cohen, sincxpress.com
%

%%

% a clear MATLAB workspace is a clear mental workspace
close all; clear

%% generate data

n    = 1000;
data = randn(n,1);
% data = exp( randn(n,1)*.8 ); % log-norm distribution

% theoretical normal distribution given N
x = linspace(-4,4,10001);
theonorm = normpdf(x);
theonorm = theonorm/max(theonorm);

% plot histograms on top of each other
[yy,xx] = histcounts(data,40);
yy = yy./max(yy);
xx = (xx(1:end-1)+xx(2:end))/2;

figure(1), clf, hold on
plot(xx,yy,'linew',3)
plot(x,theonorm,'linew',3)
legend({'Empirical';'Theoretical'},'box','off')
axis off

%% create a QQ plot

zSortData  = sort(zscore(data));
sortNormal = norminv(linspace(0,1,n));

% QQ plot is theory vs reality
figure(2), clf
plot(sortNormal,zSortData,'o','linew',3)

% set axes to be equal
lims = unique([get(gca,'xlim') get(gca,'ylim')]);
lims = [min(lims) max(lims)];
axis([lims lims]);

% draw comparison line
hold on
plot(lims,lims)

axis square
xlabel('Theoretical normal')
ylabel('Observed data')
title('QQ plot')

%% MATLAB solution

figure(3), clf
qqplot(data)
axis square

%% done.
